### PR TITLE
Fix ldap user sync missed email in email_address table (#18786)

### PR DESCRIPTION
**Pull request by lunny**
backport #18786, #18878

* Fix ldap user sync missed email in email_address table
* Fix test

### CODE REVIEW

File: models/user/user.go [hunk 1]
```diff
@@ -860,15 +861,34 @@ func updateUser(ctx context.Context, u *User, changePrimaryEmail bool) error {
...
 }
 
 // UpdateUser updates user's information.
-func UpdateUser(u *User, emailChanged bool) error {
-	return updateUser(db.DefaultContext, u, emailChanged)
+func UpdateUser(u *User, emailChanged bool, cols ...string) error {
+	return updateUser(db.DefaultContext, u, emailChanged, cols...)
 }
 
 // UpdateUserCols update user according special columns
```
Comment (delvh) [inline models/user/user.go:895]: What I just noticed (not too relevant for this branch, but could be used for main):
Isn't `updateUserCols(ctx, u, cols)` now exactly the same as `updateUser(ctx, u, false, cols)`?
The only potential difference is that `updateUser` does not allow to set the engine used, and that `updateUser` will validate the user email, and add it to the db if not present.
Should we remove `updateUserCols` then?

File: models/user/user.go [hunk 2]
```diff
@@ -860,15 +861,34 @@ func updateUser(ctx context.Context, u *User, changePrimaryEmail bool) error {
...
+		}
+
+		if !primaryEmailExist {
+			_, err = e.Insert(&EmailAddress{
+				Email:       u.Email,
+				UID:         u.ID,
+				IsActivated: true,
+				IsPrimary:   true,
+			})
+			return err
```
Comment (delvh) [inline models/user/user.go]: I think this `return err` is unnecessary, and leads to strange results:
If you call updateUser for a User whose Email didn't exist, the method will create the email in the db, not return an error,  and not update the user in the end. Is that really intended?

Comment (delvh) [reply models/user/user.go]: What I agree with would be
```go
if err != nil {
  return err
}
```

File: models/user/user.go (hunk 1 above)
Comment (lunny) [reply models/user/user.go:895]: We can remove it in main.

**Review by delvh — APPROVED** [review]

**Review by richmahn — APPROVED** [review]

**Pull request merged**